src/App.tsx

```diff
@@ -1,24 +1,258 @@
+
+    case FilterBy.COMPLETED:
+      return todos.filter(todo => todo.completed);
+
+    default:
+      return todos;
   }
+}
+
+export interface TodoItem {
+  todo: Todo,
+  isLoading: boolean,
+}
+
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodos, setTempTodos] = useState<Todo | null>(null);
+  const [activeInput, setActiveInput] = useState(true);
+  const [isQuery, setIsQuery] = useState(false);
+  const [deletedTodoIds, setDeletedTodoIds] = useState<number[]>([]);
+
+  const addError = (message: string) => {
+    setErrorMessage(message);
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  };
+
+  const loadAllTodos = async () => {
+    try {
+      const todos = await getTodos(USER_ID);
+
+      setAllTodos(todos);
+      setIsQuery(true);
+    } catch {
+      addError('Unable to update a todo');
+    } finally {
+      setTempTodos(null);
+      setActiveInput(true);
+    }
+  };
+
+  const handleRemoveTodo = useCallback(
+    (id: number) => {
+      setDeletedTodoIds(prev => [...prev, id]);
+
+      setTimeout(() => {
+        try {
+          removeTodo(id)
+            .then(() => {
+              const todosWithoutDeleted = allTodos.filter((todo) => (
+                todo.id !== id
+              ));
+
+              setAllTodos(todosWithoutDeleted);
+            });
+        } catch {
+          setErrorMessage('Unable to delete a todo');
+        } finally {
+          setDeletedTodoIds([]);
+        }
+      }, 300);
+    }, [allTodos],
+  );
+
+  useEffect(() => {
+    loadAllTodos();
+  }, []);
+
+  const handleRemoveAllComplitedTodo = useCallback(() => {
+    const completedTodos = allTodos.filter(({ completed }) => completed);
+
+    completedTodos.forEach(todo => {
+      setDeletedTodoIds(prev => [...prev, todo.id]);
+
+      setTimeout(() => {
+        try {
+          removeTodo(todo.id)
+            .then(() => {
+              const todosWithoutDeleted = allTodos.filter(({ completed }) => (
+                !completed
+              ));
+
+              setAllTodos(todosWithoutDeleted);
+            });
+        } catch {
+          setErrorMessage('Unable to delete a todo');
+        } finally {
+          setDeletedTodoIds([]);
+        }
+      }, 300);
+    });
+  }, [allTodos]);
+
+  const visibleTodos = useMemo(() => {
+    return filterTodosByCompleted(allTodos, filterType);
+  }, [allTodos, filterType]);
+
+  const countActiveTodos = useMemo(() => {
+    return visibleTodos.reduce((sum, todo) => {
+      if (!todo.completed) {
+        return sum + 1;
+      }
+
+      return sum;
+    }, 0);
+  }, [visibleTodos]);
+
+  const isComplitedTodo = allTodos.some(todo => todo.completed);
+  const isActiveTodo = allTodos.some(todo => !todo.completed);
```

Comment:
I gues this is working fine but why do unnesessery array operations?
```suggestion
  const isComplitedTodo = allTodos.length - countActiveTodos > 0;
  const isActiveTodo = countActiveTodos > 0;
```